README.rst

```diff
@@ -22,46 +22,96 @@ A pure Python micro-framework supporting seamless lazy and concurrent evaluation
-topic.
 
+The few lines above fully instantiate a computation graph, here in its simplest form with just one equation relating ``x``, ``y`` and ``s`` via the function
+``add``. Given values for the input variables ``x`` and ``y``, the value of ``s`` is resolved as follows:
 
-Getting started
-===============
+>>> pg.evaluate([s], {x: 5, y: 10})
+[15]
 
-Computation graphs are `bipartite graphs <https://en.wikipedia.org/wiki/Bipartite_graph>`_, where vertices of a one type, the variables (of type
-Variable), are connected together exclusively through vertices of another type, the operations (or simply ops, of type Op).
 
-In *paragraph*, turning a regular Python function into an op is as simple as decorating it:
+Key features
+''''''''''''
 
->>> @op
-... def f(a, b):
-...     return a * b
+The main benefits of using paragraph stem from the following features of ``pg.session.evaluate``:
 
-The operation can then be applied to objects of both Variable and non-Variable type as follows:
+Lazy evaluation
+  Irrespective of the size of the computation graph, only the operations required to evaluate the output variables are executed. Consider the following
+  extension of the above graph:
 
->>> v1 = Variable("v1")
->>> v2 = f(2, v1)
+  >>> z = pg.Variable("z")
+  >>> t = add.op(y, z)
 
-Ops differ from regular Python functions in their behavior upon receiving an argument of type Variable. In such a case, they are not executed,
-but instead pack the knowledge required for deferred execution into an instance of Variable, and return this variable.
-Then, a variable can be evaluated by invoking the function `paragraph.session.evaluate` and passing in initialization values for the input
-variables alongside the target variable:
+  Then the statement:
 
->>> value = evaluate([v2], args={v1: 4})
+  >>> pg.evaluate([t], {y: 10, z: 50})
+  [60]
 
+  just ignores the variables ``s`` and ``x`` altogether, since they do not contribute to the evaluation of ``t``. In particular, the operation ``add(x, y)``
+  is not executed.
 
-Going further
-=============
+
+Eager currying
+  Invoking an op with invariable arguments (that is, arguments that are not of type ``Variable``) just returns an invariable value: evaluation is
+  eager whenever possible. If only a subset of inputs is a variable, say ``x_p``, the computation graph can be simplified using ``solve``, which returns a new
+  variable:
+  
+  >>> x_p = Variable("x_p")
+  >>> u = add.op(s, t)
+  >>> u_xp = pg.solve([u], {x: x_p, y: 10, z: 50})[0]
+  
+  Here, ``u_xp`` is a different variable from ``u``: it depends on a single input variable (``x_p``), and it knows nothing about a variable ``y`` or ``z``,
+  instead storing a reference to the value of their sum ``t``, i.e. ``60``.
+
+  Thus, ``pg.session.solve`` acts much as ``functools.partial``, except it simplifies the system of equations where possible by executing dependent
```

Comment:
Superfluous 'dependent'

Reply:
Not really superfluous, it is really executing _only_ dependent variables.